src/Cardano/Wallet/Api/Types.hs

```diff
@@ -202,91 +199,111 @@ getApiMnemonicT (ApiMnemonicT (pw, _)) = pw
 -------------------------------------------------------------------------------}
 
-newtype DecodeApiAddressError = DecodeApiAddressError String
-    deriving stock Show
-    deriving newtype Buildable
+-- | Defines a textual encoding for a type defined within the API.
+class ToText a where
+    -- | Encode the specified value as text.
+    toText :: a -> Text
 
--- | Constructs an address from a Base58-encoded string.
---
--- Fails if the specified string is not Base58 encoded.
-decodeApiAddress :: Text -> Either DecodeApiAddressError (ApiT Address)
-decodeApiAddress x = maybe
-    (Left $ DecodeApiAddressError
-        "Unable to decode Address: expected Base58 encoding")
-    (pure . ApiT . Address)
-    (decodeBase58 bitcoinAlphabet $ T.encodeUtf8 x)
-
--- | Converts an address to a Base58-encoded string.
-encodeApiAddress :: ApiT Address -> Text
-encodeApiAddress =
-    T.decodeUtf8 . encodeBase58 bitcoinAlphabet . getAddress . getApiT
-
-newtype DecodeApiEncryptionPassphraseError
-    = DecodeApiEncryptionPassphraseError String
-    deriving stock Show
+-- | Defines a textual decoding for a type defined within the API.
+class FromText a where
+    -- | Decode the specified text as a value.
+    fromText :: Text -> Either TextDecodingError a
+
+-- | Indicates an error that occurred while decoding from text.
+newtype TextDecodingError = TextDecodingError
+    { getTextDecodingError :: String }
+    deriving stock (Eq, Show)
     deriving newtype Buildable
 
-decodeApiEncryptionPassphrase
-    :: Text
-    -> Either DecodeApiEncryptionPassphraseError
-        (ApiT (Passphrase "encryption"))
-decodeApiEncryptionPassphrase t
-    | T.length t < passphraseMinLength =
-        Left $ DecodeApiEncryptionPassphraseError $
-            "passphrase is too short: expected at least "
-            <> show passphraseMinLength <> " chars"
-    | T.length t > passphraseMaxLength =
-        Left $ DecodeApiEncryptionPassphraseError $
-            "passphrase is too long: expected at most "
-            <> show passphraseMaxLength <> " chars"
-    | otherwise =
-        pure $ ApiT $ Passphrase $ BA.convert $ T.encodeUtf8 t
-
-encodeApiEncryptionPassphrase :: ApiT (Passphrase "encryption") -> Text
-encodeApiEncryptionPassphrase (ApiT (Passphrase bytes)) =
-    T.decodeUtf8 $ BA.convert bytes
+instance FromText (ApiT Address) where
+    -- | Constructs an address from a Base58-encoded string.
+    fromText x = maybe
+        (Left $ TextDecodingError
+            "Unable to decode Address: expected Base58 encoding")
+        (pure . ApiT . Address)
+        (decodeBase58 bitcoinAlphabet $ T.encodeUtf8 x)
+
+instance ToText (ApiT Address) where
+    -- | Converts an address to a Base58-encoded string.
+    toText = T.decodeUtf8 . encodeBase58 bitcoinAlphabet . getAddress . getApiT
+
+instance FromText (ApiT AddressPoolGap) where
+    fromText t = maybe
+        (Left err)
+        (fmap ApiT . first (const err) . mkAddressPoolGap)
```

Comment:
minor note here, but in general, I'd recommend to go and pattern match here, even if there's only one case. This creates code that is slightly more resilient to the future when new branches get added to the underlying type. This way, you get a compiler error for not handling a particular case for which, here, you may want to provide a different message. 
Using `const err`, this will go unnoticed and we may end up serving a wrong error message.